walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,53 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> conflicts = getConflictingEvents(events, request);
+    return getPossibleTimeRanges(conflicts, request.getDuration());
+  }
+
+  private ArrayList<TimeRange> getConflictingEvents(
+      Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> conflicts = new ArrayList<TimeRange>();
+    for (Event event : events) {
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        conflicts.add(event.getWhen());
+      }
+    }
+    Collections.sort(conflicts, TimeRange.ORDER_BY_START);
+    return conflicts;
+  }
+
+  private ArrayList<TimeRange> getPossibleTimeRanges(
+      Collection<TimeRange> conflicts, long duration) {
+    int start = TimeRange.START_OF_DAY;
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    // These are the situations that could occur while looking for a good time
+    // s = start
+    // Case 1: s |---|  -> |--s
+    //      -> distance could be (a) large enough to add as a possible time
+    //                        or (b) not large enough
+    // Case 2: |-s-|    -> |--s
+    //
+    // Case 3: |---| s  -> |---| s
+    //
+    for (TimeRange conflict : conflicts) {
+      int end = conflict.start();
+      if (end - start >= duration) { // Case 1(a)
+        possibleTimes.add(TimeRange.fromStartEnd(start, end, false));
+      }
+      if (start < conflict.end()) { // !Case 3
+        start = conflict.end();
+      }
+    }
+    if (TimeRange.END_OF_DAY - start >= duration) {
+      possibleTimes.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+    return possibleTimes;
```

Comment:
These methods should have corresponding tests. In fact, in test driven development we write unit tests before we write functions or methods. For a more in-depth explanation see this 5-minute youtube video: https://www.youtube.com/watch?v=QCif_-r8eK4.

Reply:
Ohh wow. Yeah that makes a lot of sense. Would I have to make really extensive tests given that I use them as helpers and would I have to make them public?

Reply:
I thought it would be okay because they get tested with the main function

Reply:
You can access private methods from unit tests using reflection: https://medium.com/mindorks/how-to-unit-test-private-methods-in-java-and-kotlin-d3cae49dccd. I think you should test these methods even though they will also get exercised in the main function. That way if something breaks in one of your private methods it'll be easier to find because the test of that private method will fail. If you only test the main function and something breaks you'll have to search through all of the untested private methods to find the bug.

Reply:
Will do!

Reply:
Also, the tests shouldn't be that extensive since the logic in these methods isn't that complex. Just write one or two simple sanity checks